### PR TITLE
Fix maven conditional publishing snippets

### PR DESCRIPTION
**Pull request by abstratt**
(and corresponding sample test)

Fixes: #21640

Part of #21027

### CODE REVIEW

File: subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/SamplesMavenPublishIntegrationTest.groovy [hunk 1]
```diff
@@ -205,9 +194,6 @@ class SamplesMavenPublishIntegrationTest extends AbstractSampleIntegrationTest {
     }
 
     @UsesSample("maven-publish/publish-artifact")
```
Comment (abstratt) [inline subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/SamplesMavenPublishIntegrationTest.groovy:196]: drive-by fix

**Comment by abstratt** [comment]
@bot-gradle test BD

File: subprojects/docs/src/snippets/maven-publish/publish-artifact/kotlin/build.gradle.kts [hunk 2]
```diff
@@ -15,6 +15,7 @@ val rpmArtifact = artifacts.add("archives", rpmFile.get().asFile) {
 // end::custom-artifact[]
 
 tasks.register("rpm") {
+    val rpmFile = rpmFile
```
Comment (abstratt) [inline subprojects/docs/src/snippets/maven-publish/publish-artifact/kotlin/build.gradle.kts:18]: drive-by fix

**Comment by abstratt** [comment]
@bot-gradle help

**Comment by abstratt** [comment]
@bot-gradle test RFN

File: subprojects/docs/src/snippets/maven-publish/conditional-publishing/kotlin/build.gradle.kts [hunk 3]
```diff
@@ -38,17 +38,19 @@ publishing {
...
 // tag::task-config[]
 tasks.withType<PublishToMavenRepository>().configureEach {
-    onlyIf("publishing binary to the external repository, or binary and sources to the internal one") {
+    val check = providers.provider {
         (repository == publishing.repositories["external"] &&
             publication == publishing.publications["binary"]) ||
         (repository == publishing.repositories["internal"] &&
             publication == publishing.publications["binaryAndSources"])
     }
+    onlyIf("publishing binary to the external repository, or binary and sources to the internal one") { check.get() }
```
Comment (bamboo) [inline subprojects/docs/src/snippets/maven-publish/conditional-publishing/kotlin/build.gradle.kts:47]: 💄 I'd keep the block syntax as the Groovy version.

File: subprojects/docs/src/snippets/maven-publish/conditional-publishing/kotlin/build.gradle.kts [hunk 4]
```diff
@@ -38,17 +38,19 @@ publishing {
...
 }
 tasks.withType<PublishToMavenLocal>().configureEach {
-    onlyIf("publishing binary and sources") {
+    val check = providers.provider {
         publication == publishing.publications["binaryAndSources"]
     }
+    onlyIf("publishing binary and sources") { check.get() }
```
Comment (bamboo) [inline subprojects/docs/src/snippets/maven-publish/conditional-publishing/kotlin/build.gradle.kts:53]: 💄 Same as above.

File: subprojects/docs/src/snippets/maven-publish/publish-artifact/kotlin/build.gradle.kts (hunk 2 above)
Comment (bamboo) [reply subprojects/docs/src/snippets/maven-publish/publish-artifact/kotlin/build.gradle.kts:18]: 💄 A clarifying comment could help:

```kotlin
val rpmFile = rpmFile // Reduce scope of property for compatibility with the configuration cache
```

**Comment by abstratt** [comment]
@bot-gradle test BD

File: subprojects/docs/src/snippets/maven-publish/conditional-publishing/groovy/build.gradle [hunk 5]
```diff
@@ -38,17 +38,23 @@ publishing {
 
 // tag::task-config[]
 tasks.withType(PublishToMavenRepository) {
-    onlyIf("publishing binary to the external repository, or binary and sources to the internal one") {
+    def check = providers.provider {
```
Comment (bamboo) [inline subprojects/docs/src/snippets/maven-publish/conditional-publishing/groovy/build.gradle:41]: 💄 Perhaps `spec` since that's the term used in the `onlyIf` documentation?

It would be nice to have an explanation comment as well:
```
    // Capture spec result in the configuration cache
    def spec = ...
```

Comment (abstratt) [reply subprojects/docs/src/snippets/maven-publish/conditional-publishing/groovy/build.gradle:41]: I tend to avoid using variable/parameter names that describe the type, as that is what type declarations are for, and prefer to name after role instead.

I will go with `checkSpec` as a middle ground - stop me if that bothers you.

Comment (bamboo) [reply subprojects/docs/src/snippets/maven-publish/conditional-publishing/groovy/build.gradle:41]: The docs also use the term `predicate`.

Comment (bamboo) [reply subprojects/docs/src/snippets/maven-publish/conditional-publishing/groovy/build.gradle:41]: And by the way, `spec` was not about the type. `spec` is also common noun used in the docs.

Comment (abstratt) [reply subprojects/docs/src/snippets/maven-publish/conditional-publishing/groovy/build.gradle]: Cool, went with `predicate`, as a more common name

**Comment by abstratt** [comment]
@bot-gradle test this

File: subprojects/docs/src/snippets/maven-publish/conditional-publishing/groovy/build.gradle [hunk 6]
```diff
@@ -38,17 +38,23 @@ publishing {
 
 // tag::task-config[]
 tasks.withType(PublishToMavenRepository) {
-    onlyIf("publishing binary to the external repository, or binary and sources to the internal one") {
+    def predicate = providers.provider {
```
Comment (bamboo) [inline subprojects/docs/src/snippets/maven-publish/conditional-publishing/groovy/build.gradle:41]: 💄 Just `provider` instead of `providers.provider` like in the other PR?

**Review by bamboo — APPROVED** [review]
👍

**Comment by abstratt** [comment]
@bot-gradle test and merge

**Comment by abstratt** [comment]
@bot-gradle cancel

**Comment by abstratt** [comment]
@bot-gradle test and merge
